File: code/nodejs/07-harness/pii_benchmark.ts
/**
 * PII detection accuracy benchmark.
 *
 * Tests PII detection patterns against labelled examples and reports
 * per-type precision, recall, and F1.
 * See: docs/07-harness-engineering/02-input-guardrails-and-validation.md
 */

// ---------------------------------------------------------------------------
// PII detector (self-contained — mirrors input_guardrail_pipeline)
// ---------------------------------------------------------------------------

export type PIIType = "email" | "phone" | "ssn" | "credit_card" | "api_key";

export interface PIIMatch {
  type: PIIType;
  value: string;
  start: number;
  end: number;
}

const PII_PATTERNS: Record<PIIType, RegExp> = {
  email:       /\b[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}\b/g,
  phone:       /\b(?:\+1[-.\s]?)?\(?\d{3}\)?[-.\s]\d{3}[-.\s]\d{4}\b/g,
  ssn:         /\b\d{3}-\d{2}-\d{4}\b/g,
  credit_card: /\b(?:\d[ -]?){13,16}\b/g,
  api_key:     /\b(?:sk-[a-zA-Z0-9]{20,}|AIza[a-zA-Z0-9_\-]{35}|AKIA[A-Z0-9]{16})\b/g,
};

export function detectPII(text: string): PIIMatch[] {
  const matches: PIIMatch[] = [];
  for (const [type, pattern] of Object.entries(PII_PATTERNS) as Array<[PIIType, RegExp]>) {
    const re = new RegExp(pattern.source, pattern.flags);
    let m: RegExpExecArray | null;
    while ((m = re.exec(text)) !== null) {
      matches.push({ type, value: m[0], start: m.index, end: m.index + m[0].length });
    }
  }
  return matches;
}

// ---------------------------------------------------------------------------
// Test cases
// ---------------------------------------------------------------------------

export interface BenchmarkCase {
  text: string;
  expectedPII: boolean;
  expectedTypes: PIIType[];
  description: string;
}

const TRUE_POSITIVES: BenchmarkCase[] = [
  { text: "Email me at alice@example.com", expectedPII: true, expectedTypes: ["email"], description: "simple email" },
  { text: "Call 555-867-5309", expectedPII: true, expectedTypes: ["phone"], description: "US phone" },
  { text: "SSN is 123-45-6789", expectedPII: true, expectedTypes: ["ssn"], description: "SSN" },
  { text: "API key: sk-abc123def456ghi789jkl012mno345pqr678stu", expectedPII: true, expectedTypes: ["api_key"], description: "OpenAI-style key" },
  { text: "My email alice@example.com and phone 415.555.2671", expectedPII: true, expectedTypes: ["email", "phone"], description: "multiple PII" },
];

const TRUE_NEGATIVES: BenchmarkCase[] = [
  { text: "The quick brown fox jumps over the lazy dog", expectedPII: false, expectedTypes: [], description: "no PII" },
  { text: "Version 1.2.3 released on 2024-01-01", expectedPII: false, expectedTypes: [], description: "version numbers" },
  { text: "Connect to localhost:8080", expectedPII: false, expectedTypes: [], description: "localhost URL" },
  { text: "HTTP 200 OK response", expectedPII: false, expectedTypes: [], description: "HTTP status" },
  { text: "Array index: arr[123]", expectedPII: false, expectedTypes: [], description: "array index" },
];

// ---------------------------------------------------------------------------
// Benchmark runner
// ---------------------------------------------------------------------------

export interface TypeMetrics {
  truePositives: number;
  falsePositives: number;
  falseNegatives: number;
  precision: number;
  recall: number;
  f1: number;
}

export interface BenchmarkReport {
  totalCases: number;
  overallAccuracy: number;
  typeMetrics: Record<PIIType, TypeMetrics>;
  falsePositives: Array<{ text: string; detectedTypes: PIIType[] }>;
  falseNegatives: Array<{ text: string; expectedTypes: PIIType[] }>;
}

export function runBenchmark(cases: BenchmarkCase[] = [...TRUE_POSITIVES, ...TRUE_NEGATIVES]): BenchmarkReport {
  let correct = 0;
  const tpCount: Record<PIIType, number> = { email: 0, phone: 0, ssn: 0, credit_card: 0, api_key: 0 };
  const fpCount: Record<PIIType, number> = { email: 0, phone: 0, ssn: 0, credit_card: 0, api_key: 0 };
  const fnCount: Record<PIIType, number> = { email: 0, phone: 0, ssn: 0, credit_card: 0, api_key: 0 };
  const falsePositives: Array<{ text: string; detectedTypes: PIIType[] }> = [];
  const falseNegatives: Array<{ text: string; expectedTypes: PIIType[] }> = [];

  for (const c of cases) {
    const matches = detectPII(c.text);
    const detectedTypes = [...new Set(matches.map((m) => m.type))];
    const detectedPII = detectedTypes.length > 0;

    if (detectedPII === c.expectedPII) correct++;

    // Per-type accounting
    for (const expected of c.expectedTypes) {
      if (detectedTypes.includes(expected)) tpCount[expected]++;
      else { fnCount[expected]++; falseNegatives.push({ text: c.text, expectedTypes: c.expectedTypes }); }
    }
    for (const detected of detectedTypes) {
      if (!c.expectedTypes.includes(detected)) {
        fpCount[detected]++;
        falsePositives.push({ text: c.text, detectedTypes });
      }
    }
  }

  const typeMetrics = {} as Record<PIIType, TypeMetrics>;
  for (const t of Object.keys(PII_PATTERNS) as PIIType[]) {
    const tp = tpCount[t], fp = fpCount[t], fn = fnCount[t];
    const precision = tp + fp > 0 ? tp / (tp + fp) : 1;
    const recall = tp + fn > 0 ? tp / (tp + fn) : 1;
    const f1 = precision + recall > 0 ? 2 * precision * recall / (precision + recall) : 0;
    typeMetrics[t] = { truePositives: tp, falsePositives: fp, falseNegatives: fn, precision, recall, f1 };
  }

  return {
    totalCases: cases.length,
    overallAccuracy: correct / cases.length,
    typeMetrics,
    falsePositives: [...new Set(falsePositives.map((f) => JSON.stringify(f)))].map((s) => JSON.parse(s) as { text: string; detectedTypes: PIIType[] }),
    falseNegatives: [...new Set(falseNegatives.map((f) => JSON.stringify(f)))].map((s) => JSON.parse(s) as { text: string; expectedTypes: PIIType[] }),
  };
}

/** Print benchmark report. */
export function printBenchmarkReport(report: BenchmarkReport): void {
  console.log(`\n=== PII Detection Benchmark (${report.totalCases} cases) ===`);
  console.log(`  Overall Accuracy: ${(report.overallAccuracy * 100).toFixed(1)}%`);
  console.log("\n  Per-type metrics:");
  console.log("  Type".padEnd(14) + "Precision".padStart(10) + "Recall".padStart(10) + "F1".padStart(10));
  console.log("  " + "-".repeat(42));
  for (const [type, m] of Object.entries(report.typeMetrics)) {
    console.log(
      `  ${type.padEnd(12)}` +
      `${(m.precision * 100).toFixed(1)}%`.padStart(10) +
      `${(m.recall * 100).toFixed(1)}%`.padStart(10) +
      `${(m.f1 * 100).toFixed(1)}%`.padStart(10)
    );
  }
  if (report.falsePositives.length) {
    console.log(`\n  False positives: ${report.falsePositives.length}`);
  }
  if (report.falseNegatives.length) {
    console.log(`  False negatives: ${report.falseNegatives.length}`);
  }
}

// Demo
function main(): void {
  const report = runBenchmark();
  printBenchmarkReport(report);
}

main();

File: code/go/07-harness/pii_benchmark.go
// pii_benchmark.go — PII detection accuracy benchmark (Go port).
//
// See: docs/07-harness-engineering/02-input-guardrails-and-validation.md
package main

import (
	"fmt"
	"regexp"
	"strings"
)

// ---------------------------------------------------------------------------
// PII patterns
// ---------------------------------------------------------------------------

// PIIPattern describes a regex-based PII detector.
type PIIPattern struct {
	Name    string
	Pattern *regexp.Regexp
}

// DefaultPIIPatterns returns common PII patterns.
func DefaultPIIPatterns() []PIIPattern {
	return []PIIPattern{
		{"email", regexp.MustCompile(`(?i)[a-z0-9._%+\-]+@[a-z0-9.\-]+\.[a-z]{2,}`)},
		{"ssn", regexp.MustCompile(`\b\d{3}-\d{2}-\d{4}\b`)},
		{"credit_card", regexp.MustCompile(`\b(?:\d[ -]?){13,16}\b`)},
		{"phone_us", regexp.MustCompile(`\b(?:\+1[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b`)},
		{"ip_address", regexp.MustCompile(`\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b`)},
	}
}

// DetectPII scans text for PII using the given patterns.
// Returns a map of pattern name → slice of matches.
func DetectPII(text string, patterns []PIIPattern) map[string][]string {
	results := make(map[string][]string)
	for _, p := range patterns {
		matches := p.Pattern.FindAllString(text, -1)
		if len(matches) > 0 {
			results[p.Name] = matches
		}
	}
	return results
}

// RedactPII replaces PII matches with [REDACTED-<type>].
func RedactPII(text string, patterns []PIIPattern) string {
	for _, p := range patterns {
		text = p.Pattern.ReplaceAllStringFunc(text, func(m string) string {
			return fmt.Sprintf("[REDACTED-%s]", strings.ToUpper(p.Name))
		})
	}
	return text
}

// ---------------------------------------------------------------------------
// Benchmark types
// ---------------------------------------------------------------------------

// PIIBenchmarkCase is a single test case for PII detection.
type PIIBenchmarkCase struct {
	ID          string
	Text        string
	ExpectedPII map[string]bool // pattern name → should be detected
}

// PIIBenchmarkResult holds the result for one case.
type PIIBenchmarkResult struct {
	Case       PIIBenchmarkCase
	Detected   map[string][]string
	Correct    bool
}

// PIIBenchmark runs PII detection over a set of test cases.
type PIIBenchmark struct {
	patterns []PIIPattern
	cases    []PIIBenchmarkCase
}

// NewPIIBenchmark creates a benchmark with the default patterns.
func NewPIIBenchmark() *PIIBenchmark {
	return &PIIBenchmark{patterns: DefaultPIIPatterns()}
}

// AddCase adds a test case.
func (b *PIIBenchmark) AddCase(tc PIIBenchmarkCase) {
	b.cases = append(b.cases, tc)
}

// Run executes all test cases and returns results.
func (b *PIIBenchmark) Run() []PIIBenchmarkResult {
	results := make([]PIIBenchmarkResult, 0, len(b.cases))
	for _, tc := range b.cases {
		detected := DetectPII(tc.Text, b.patterns)
		correct := true
		for piiType, shouldDetect := range tc.ExpectedPII {
			_, wasDetected := detected[piiType]
			if wasDetected != shouldDetect {
				correct = false
			}
		}
		results = append(results, PIIBenchmarkResult{
			Case:     tc,
			Detected: detected,
			Correct:  correct,
		})
	}
	return results
}

// PrintReport prints a summary of benchmark results.
func (b *PIIBenchmark) PrintReport(results []PIIBenchmarkResult) {
	passed := 0
	for _, r := range results {
		status := "✗"
		if r.Correct {
			status = "✓"
			passed++
		}
		fmt.Printf("  %s [%s] %s\n", status, r.Case.ID, truncStr(r.Case.Text, 50))
		for piiType, matches := range r.Detected {
			fmt.Printf("      Detected %s: %v\n", piiType, matches)
		}
	}
	fmt.Printf("\nResult: %d/%d passed\n", passed, len(results))
}

// RunPIIBenchmarkDemo demonstrates PII detection and redaction.
func RunPIIBenchmarkDemo() {
	bench := NewPIIBenchmark()
	bench.AddCase(PIIBenchmarkCase{
		ID:   "B001",
		Text: "Contact me at alice@example.com or call 415-555-1234",
		ExpectedPII: map[string]bool{"email": true, "phone_us": true},
	})
	bench.AddCase(PIIBenchmarkCase{
		ID:   "B002",
		Text: "My SSN is 123-45-6789",
		ExpectedPII: map[string]bool{"ssn": true},
	})
	bench.AddCase(PIIBenchmarkCase{
		ID:   "B003",
		Text: "The weather is sunny today",
		ExpectedPII: map[string]bool{},
	})

	results := bench.Run()
	bench.PrintReport(results)

	// Demo redaction
	text := "Please contact john@acme.com, SSN 987-65-4321"
	redacted := RedactPII(text, DefaultPIIPatterns())
	fmt.Printf("\nOriginal: %s\nRedacted: %s\n", text, redacted)
}
